# Data Source - Coffee Shop Sales - [Kaggle](https://www.kaggle.com/datasets/ahmedabbas757/coffee-sales)

In [ ]:
import pandas as pd

In [11]:
sales = pd.read_excel("../data/raw/Coffee Shop Sales.xlsx")
sales.head(10)

transaction_id transaction_date transaction_time  transaction_qty  \
0               1       2023-01-01         07:06:11                2   
1               2       2023-01-01         07:08:56                2   
2               3       2023-01-01         07:14:04                2   
3               4       2023-01-01         07:20:24                1   
4               5       2023-01-01         07:22:41                2   
5               6       2023-01-01         07:22:41                1   
6               7       2023-01-01         07:25:49                1   
7               8       2023-01-01         07:33:34                2   
8               9       2023-01-01         07:39:13                1   
9              10       2023-01-01         07:39:34                2   

   store_id   store_location  product_id  unit_price    product_category  \
0         5  Lower Manhattan          32        3.00              Coffee   
1         5  Lower Manhattan          57        3.10                 Tea   
2         5  Lower Manhattan          59        4.50  Drinking Chocolate   
3         5  Lower Manhattan          22        2.00              Coffee   
4         5  Lower Manhattan          57        3.10                 Tea   
5         5  Lower Manhattan          77        3.00              Bakery   
6         5  Lower Manhattan          22        2.00              Coffee   
7         5  Lower Manhattan          28        2.00              Coffee   
8         5  Lower Manhattan          39        4.25              Coffee   
9         5  Lower Manhattan          58        3.50  Drinking Chocolate   

            product_type               product_detail  
0  Gourmet brewed coffee                  Ethiopia Rg  
1        Brewed Chai tea     Spicy Eye Opener Chai Lg  
2          Hot chocolate            Dark chocolate Lg  
3            Drip coffee  Our Old Time Diner Blend Sm  
4        Brewed Chai tea     Spicy Eye Opener Chai Lg  
5                  Scone                Oatmeal Scone  
6            Drip coffee  Our Old Time Diner Blend Sm  
7  Gourmet brewed coffee    Columbian Medium Roast Sm  
8       Barista Espresso                     Latte Rg  
9          Hot chocolate            Dark chocolate Rg

In [10]:
sales.tail(10)

transaction_id transaction_date transaction_time  transaction_qty  \
149106          149447       2023-06-30         19:56:45                1   
149107          149448       2023-06-30         19:57:49                2   
149108          149449       2023-06-30         19:57:49                1   
149109          149450       2023-06-30         20:08:38                2   
149110          149451       2023-06-30         20:16:31                2   
149111          149452       2023-06-30         20:18:41                2   
149112          149453       2023-06-30         20:25:10                2   
149113          149454       2023-06-30         20:31:34                1   
149114          149455       2023-06-30         20:57:19                1   
149115          149456       2023-06-30         20:57:19                2   

        store_id   store_location  product_id  unit_price product_category  \
149106         8   Hell's Kitchen          33        3.50           Coffee   
149107         3          Astoria          54        2.50              Tea   
149108         3          Astoria          70        3.25           Bakery   
149109         8   Hell's Kitchen          55        4.00              Tea   
149110         5  Lower Manhattan          42        2.50              Tea   
149111         8   Hell's Kitchen          44        2.50              Tea   
149112         8   Hell's Kitchen          49        3.00              Tea   
149113         8   Hell's Kitchen          45        3.00              Tea   
149114         8   Hell's Kitchen          40        3.75           Coffee   
149115         8   Hell's Kitchen          64        0.80         Flavours   

                 product_type           product_detail  
149106  Gourmet brewed coffee              Ethiopia Lg  
149107        Brewed Chai tea  Morning Sunrise Chai Rg  
149108                  Scone          Cranberry Scone  
149109        Brewed Chai tea  Morning Sunrise Chai Lg  
149110      Brewed herbal tea           Lemon Grass Rg  
149111      Brewed herbal tea            Peppermint Rg  
149112       Brewed Black tea     English Breakfast Lg  
149113      Brewed herbal tea            Peppermint Lg  
149114       Barista Espresso               Cappuccino  
149115          Regular syrup           Hazelnut syrup

In [8]:
sales.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 149116 entries, 0 to 149115
Data columns (total 11 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   transaction_id    149116 non-null  int64         
 1   transaction_date  149116 non-null  datetime64[ns]
 2   transaction_time  149116 non-null  object        
 3   transaction_qty   149116 non-null  int64         
 4   store_id          149116 non-null  int64         
 5   store_location    149116 non-null  object        
 6   product_id        149116 non-null  int64         
 7   unit_price        149116 non-null  float64       
 8   product_category  149116 non-null  object        
 9   product_type      149116 non-null  object        
 10  product_detail    149116 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(4), object(5)
memory usage: 12.5+ MB


In [9]:
sales.describe()

transaction_id               transaction_date  transaction_qty  \
count   149116.000000                         149116    149116.000000   
mean     74737.371872  2023-04-15 11:50:32.173609984         1.438276   
min          1.000000            2023-01-01 00:00:00         1.000000   
25%      37335.750000            2023-03-06 00:00:00         1.000000   
50%      74727.500000            2023-04-24 00:00:00         1.000000   
75%     112094.250000            2023-05-30 00:00:00         2.000000   
max     149456.000000            2023-06-30 00:00:00         8.000000   
std      43153.600016                            NaN         0.542509   

            store_id     product_id     unit_price  
count  149116.000000  149116.000000  149116.000000  
mean        5.342063      47.918607       3.382219  
min         3.000000       1.000000       0.800000  
25%         3.000000      33.000000       2.500000  
50%         5.000000      47.000000       3.000000  
75%         8.000000      60.000000       3.750000  
max         8.000000      87.000000      45.000000  
std         2.074241      17.930020       2.658723

The data set represents a **synthetic** one where **149,115** transactions in **coffee shops** of different locations were recorded accross diffent times of the day. The data set was conceptualized to not contain missing values.

For the purposes of this **Association Rule Analysis** experiment, only the `transaction_date`, `transaction_time`, `transaction_qty`, `store_id` and `product_type` will be kept.
`transaction_id` is not significant in the study as an ID represents a single item and not an actual transaction in the sense of a **basket**.

In [33]:
sales_selected = sales.loc[:, ["transaction_date", "transaction_time", "transaction_qty", "store_id", "product_type"]]
sales_selected

transaction_date transaction_time  transaction_qty  store_id  \
0            2023-01-01         07:06:11                2         5   
1            2023-01-01         07:08:56                2         5   
2            2023-01-01         07:14:04                2         5   
3            2023-01-01         07:20:24                1         5   
4            2023-01-01         07:22:41                2         5   
...                 ...              ...              ...       ...   
149111       2023-06-30         20:18:41                2         8   
149112       2023-06-30         20:25:10                2         8   
149113       2023-06-30         20:31:34                1         8   
149114       2023-06-30         20:57:19                1         8   
149115       2023-06-30         20:57:19                2         8   

                 product_type  
0       Gourmet brewed coffee  
1             Brewed Chai tea  
2               Hot chocolate  
3                 Drip coffee  
4             Brewed Chai tea  
...                       ...  
149111      Brewed herbal tea  
149112       Brewed Black tea  
149113      Brewed herbal tea  
149114       Barista Espresso  
149115          Regular syrup  

[149116 rows x 5 columns]

`product_category` and `product_detail` could be used as well for the **Association Rule analysis**, but we preferred using `product_type` as it's closer to what day-to-day orders are called.

`transaction_date`, `transaction_time` and `store_id` will be used to spot individual basket items as those will have identical values for the previously stated features.

In [31]:
product_types = set(sales["product_type"])
repr(product_types)

"{'Organic Beans', 'Brewed Black tea', 'Brewed herbal tea', 'Green tea', 'Organic Chocolate', 'Scone', 'Biscotti', 'Premium brewed coffee', 'Sugar free syrup', 'Black tea', 'Brewed Chai tea', 'Gourmet brewed coffee', 'Pastry', 'Drip coffee', 'Brewed Green tea', 'Hot chocolate', 'Organic brewed coffee', 'Green beans', 'Herbal tea', 'House blend Beans', 'Premium Beans', 'Espresso Beans', 'Gourmet Beans', 'Housewares', 'Clothing', 'Drinking Chocolate', 'Chai tea', 'Regular syrup', 'Barista Espresso'}"

In [32]:
print(f"Overall number of distinct product types: {len(product_types)}")

Overall number of distinct product types: 29


#### Exporting data

In [ ]:
sales_selected.to_csv("../data/interim/selected_sales_features.csv", index=False)

: 